src/hugr/views/sibling.rs

```diff
@@ -0,0 +1,572 @@
+    /// documentation.
+    pub fn from_boundary_edges_with_checker(
+        base: &'g Base,
+        incoming: impl IntoIterator<Item = (Node, Port)>,
+        outgoing: impl IntoIterator<Item = (Node, Port)>,
+        checker: &mut ConvexChecker<&'g Base::Portgraph>,
+    ) -> Result<Self, InvalidSubgraph>
+    where
+        Base: HugrView,
+    {
+        let pg = base.portgraph();
+        let to_pg = |(n, p): (Node, Port)| pg.port_index(n.index, p.offset).expect("invalid port");
+        let incoming = incoming.into_iter().flat_map(|(n, p)| {
+            if p.direction() == Direction::Outgoing {
+                base.linked_ports(n, p).map(to_pg).collect()
+            } else {
+                vec![to_pg((n, p))]
+            }
+        });
+        let outgoing = outgoing.into_iter().flat_map(|(n, p)| {
+            if p.direction() == Direction::Incoming {
+                base.linked_ports(n, p).map(to_pg).collect()
+            } else {
+                vec![to_pg((n, p))]
+            }
+        });
+        let subpg = Subgraph::new_subgraph(pg, incoming.chain(outgoing));
+        if !subpg.is_convex_with_checker(checker) {
+            return Err(InvalidSubgraph::NotConvex);
+        }
+        let nodes = subpg.nodes_iter().map_into().collect_vec();
+        if nodes.is_empty() {
+            return Err(InvalidSubgraph::EmptySubgraph);
+        }
+        let Some(parent) = base.get_parent(nodes[0]) else {
+            return Err(InvalidSubgraph::NoSharedParent);
+        };
+        if !nodes.iter().all(|&n| base.get_parent(n) == Some(parent)) {
+            return Err(InvalidSubgraph::NoSharedParent);
+        }
+        Ok(Self { base, nodes })
+    }
+
+    /// Create a new convex sibling subgraph from a set of nodes.
+    ///
+    /// This fails if the set of nodes is not convex, nodes do not share a
+    /// common parent or the subgraph is empty.
+    pub fn try_new(base: &'g Base, nodes: Vec<Node>) -> Result<Self, InvalidSubgraph>
+    where
+        Base: HugrView,
+    {
+        let mut checker = ConvexChecker::new(base.portgraph());
+        Self::try_new_with_checker(base, nodes, &mut checker)
+    }
+
+    /// Create a new convex sibling subgraph from a set of nodes.
+    ///
+    /// Provide a [`ConvexChecker`] instance to avoid constructing one for
+    /// faster convexity check. If you do not have one, use [`SiblingSubgraph::try_new`].
+    ///
+    /// This fails if the set of nodes is not convex, nodes do not share a
+    /// common parent or the subgraph is empty.
+    pub fn try_new_with_checker(
+        base: &'g Base,
+        nodes: Vec<Node>,
+        checker: &mut ConvexChecker<&'g Base::Portgraph>,
+    ) -> Result<Self, InvalidSubgraph>
+    where
+        Base: HugrView,
+    {
+        if !checker.is_node_convex(nodes.iter().map(|n| n.index)) {
+            return Err(InvalidSubgraph::NotConvex);
+        }
+        if nodes.is_empty() {
```

Comment:
Do we need to rule this out? Since `SiblingSubgraph` is now represented as base+nodes, and the latter can be an empty vec, can we just proceed?

An empty SiblingSubgraph seems to follow naturally if you pass in an empty set of boundary edges; the case of a SiblingSubgraph being *the whole SiblingGraph* to me seems less natural. And I'm not sure I really see a SimpleReplacement replacing the input/output nodes either (@cqc-alec ?) - we'd need a special case as the replacement part thereof specifically does *not* copy its input/output nodes in.

Reply:
The problem with the empty case is that there is no way to specify _where_ this subgraph lives when using nodes to define it. It could be on any set of edges... Furthermore, from a pattern matching perspective, empty subgraphs match everywhere, which blows up the number of matches and makes them useless.

If you wanted to be able to refer to empty subgraphs, you'd have to revert to a more general representation of subgraph using boundary edges as I was suggesting initially. 

Finally, replacing entire graphs is useful for testing and could also apply to non-DFG graphs. I cannot tell you whether there is a use case for them presently, but I see no harm in allowing it.

Reply:
A more grounded objection (and the actual reason this check was added):

This constructor with `(base, nodes)` signature cannot infer the shared parent if `nodes` is empty. What shall we do in that case then? A `SiblingSubgraph` should always have a parent.

Reply:
Yes, ok

Reply:
Sorry for noise!